[ADD] estate: add real estate module (framework 101 tutorial)

robal-odoo opened the pull request.

robal-odoo committed:
[ADD] estate: add new real estate module

Framework 101 tutorial
- add estate module
- add estate_account module for integration with invoicing

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1077.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1077)

ushyme commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'name': 'Real Estate',
+    'description': 'Real Estate Advertising',
+    # 'version': '1.0',
+    'author': 'robal',
```
```suggestion
    'author': 'Odoo S.A.',
```
As an Odoo employee, you are creating this module in the name of Odoo

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+    name = fields.Char(string="Title", required=True)
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date(string="Available From", default=fields.Date.today() + relativedelta(months=3), copy=False)
```
```suggestion
    date_availability = fields.Date(string="Available From", default=lambda self: fields.Date.context_today(self) + relativedelta(months=3), copy=False)
```


`date_availability`'s default was calculated only once at server launch. Wrapping it in a `lambda` fixes this by computing it dynamically for each new record. I also used `fields.Date.context_today(self)` for timezone correctness.

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+        for record in self:
+            if record.state == "sold":
+                error_msg = "This property has already been sold."
+                raise UserError(error_msg)
```
```suggestion
            if record.state == "sold":
                continue
```
In the `action_` methods, it's slightly more user-friendly to silently ignore a button click if the record is already in the target state, rather than raising an error.

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+                raise UserError(error_msg)
+
+            if record.state == "cancelled":
+                error_msg = "Cancelled properties cannot be sold."
```
```suggestion
                error_msg = record.env._("Cancelled properties cannot be sold.")
```
User-facing strings in errors and constraints must be wrapped in `record.env._()` to be translatable.

robal-odoo replied:
Is there a practical difference between using `record.env._()` and using `_()` imported with `from odoo import _`  or the one from `self.env._()` ?

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,68 @@
+
+
+def _get_reference_date(offer: EstatePropertyOffer) -> date:
+    return fields.Date.today() if not offer.create_date else offer.create_date.date()
```
I don't know if this function is really helpful here. I think the code would overall be more readable if this logic was inlined directly in the compute/inverse method.

robal-odoo replied:
Should I use `context_today` like in the other model ?

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,75 @@
+        for val in vals:
+            estate_property = self.env["estate.property"].browse(val["property_id"])
+            best_price = estate_property.best_price or 0.0
+            if val["price"] < best_price:
```
You should also use `float_compare()` here.

ushyme commented on estate_account/__manifest__.py:
```diff
@@ -0,0 +1,8 @@
+{
+    "name": "Real Estate Accounting",
+    "description": "Real Estate Accounting (link module)",
+    "author": "robal",
```
```suggestion
    "author": "Odoo S.A.",
```
As an employee, you work in the name of Odoo.

ushyme commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,21 @@
+
+    def action_set_sold(self) -> bool:
+        percent_description = _("6 percent of selling price")
+        admin_fee_description = _("Administrative fees")
```
You don't need to explicitly call `_()` for field values. If the fields are marked as translatable, the system takes care of the rest.

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_new_or_cancelled(self) -> None:
+        if any(record.state not in {"new", "cancelled"} for record in self):
+            raise UserError(_("Cannot delete properties unless they are new or cancelled."))
```
```suggestion
            raise UserError(self.env._("Cannot delete properties unless they are new or cancelled."))
```
It's best to access the translation method defined on the environment object. The older standalone `_()` function has a few issues that the newer `self.env._()` doesn't.

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+        for record in self:
+            # Selling price is zero when no offer has been accepted
+            if not float_is_zero(record.selling_price, precision_digits=2) and float_compare(record.selling_price, 0.9 * record.expected_price, precision_digits=2) < 0:
+                raise ValidationError(_("A property's selling price cannot be lower that 90 percent of its expected price."))
```
```suggestion
                raise ValidationError(_("A property's selling price cannot be lower than 90 percent of its expected price."))
```
small typo

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+
+    # Methods that trigger on changes
+    @api.onchange("garden")
+    def _compute_garden_defaults(self) -> None:
```
```suggestion
    def _onchange_garden_defaults(self) -> None:
```
This is an `onchange`, not a `compute`. Should be named `_onchange_garden_defaults` or `_onchange_garden`.

ushyme commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,21 @@
+                    Command.create({"name": percent_description, "quantity": 1, "price_unit": 0.06 * record.selling_price}),
+                    Command.create({"name": admin_fee_description, "quantity": 1, "price_unit": 100.0}),
+                ],
+                }
```
```suggestion
            }
```

ushyme commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "Real Estate",
+    "description": "Real Estate Advertising",
+    "author": "robal",
```
```suggestion
    "author": "Odoo S.A.",
```
As an employee, you work in the name of Odoo.

ushyme commented on .gitignore:
```diff
@@ -127,3 +127,6 @@ dmypy.json
 .pyre/
+
+# VScode
+.vscode/
```
```suggestion
```
This is personal configuration that shouldn't be pushed to the general codebase.

ushyme reviewed (commented):
Good work so far. I just have a few suggestions.

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,72 @@
+    @api.model
+    def create(self, vals):
+        for val in vals:
+            estate_property = self.env["estate.property"].browse(val["property_id"])
```
If `vals` contains offers for multiple properties, this will trigger multiple singleton browses.
Extract all `property_ids` from `vals` first, browse them in one go, and then iterate.

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,72 @@
+    @api.model
+    def create(self, vals):
+        for val in vals:
+            estate_property = self.env["estate.property"].browse(val["property_id"])
+            best_price = estate_property.best_price or 0.0
+            if float_compare(val["price"], best_price, precision_digits=2) < 0:
+                raise UserError(self.env._("A new offer must match or exceed the price of the current best offer."))
+            estate_property.state = "offer_received"
+        return super().create(vals)
+
+    # Public methods
+    def action_accept(self) -> bool:
+        # TODO double check validation
+        for record in self:
+            if record.property_id.offer_ids.filtered(lambda r: r.status == "accepted"):
+                raise UserError(record.env._("Cannot accept this offer because another offer has already been accepted for the property."))
+
+            record.status = "accepted"
+            record.property_id.selling_price = record.price
+            record.property_id.buyer_id = record.partner_id
+            record.property_id.state = "offer_accepted"
+        return True
```
what if the property is already sold or cancelled ?

ushyme commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,21 @@
+                ],
+            }
+            self.env["account.move"].create(account_move)
+        return super().action_set_sold()
```
Currently, you are calling `self.env["account.move"].create(account_move)` **inside** the loop. This triggers an `INSERT` and potentially multiple compute/inverse calls per record.

Also, what if the `buyer_id` is not set ?

ushyme reviewed (commented):
Looks good 👍🏾; Just a few remarks.

robal-odoo committed:
[CLN] estate: fix typo in xml view

robal-odoo committed:
[FIX] estate: add extra validation when accepting an offer

- Stop user from accepting offers if the linked property is sold or cancelled

robal-odoo committed:
[FIX] estate_account: optimize invoice creation

- Move the call to "create" outside the loop

robal-odoo committed:
[FIX] estate: optimize create method on offer model

- Extract the calls to "browse" outside the loop

robal-odoo committed:
[FIX] estate: add extra validation on "sold" action on property model

- Stop properties from being marked as sold if there isn't an accepted offer

robal-odoo committed:
[FIX] estate_account: ensure buyer_id is set when selling a property

- Leverage the validation logic from the parent model
- That validation already ensures there is an accepted offer, which means that buyer_id is set

robal-odoo committed:
[FIX] estate: fix validation condition on property "sold" action

robal-odoo committed:
[FIX] estate: add extra validation on estate_property_offer actions (accept, refuse)

- Cannot change the status of an offer if it's already set
- Cannot accept/refuse offers on a property that's cancelled or sold

robal-odoo committed:
[FIX] estate: add backend validation for offer_ids
